### PR TITLE
Optimize `scan` with a lookup table to reduce branching

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
This change introduces two lookup tables to the lexical scanner/tokenizer. One is a simple array-based lookup for entities in the ASCII range, and the other is a `Map` for wider codepoints.

These tables assign categories to codepoints to indicate how they should be handled. For example, all line breaks, single-line whitespace, and single-character tokens can be handled generically in their own respective fashion. Doing so allows us to perform fewer individual checks to jump into or skip over a given branch.

One additional optimization is the way in which simple single-character tokens are handled is that the token itself is encoded in the lower 8 bits the category value. This allows us to avoid any branching for such cases.

This all comes at the expense of some memory overhead, along with some overhead from fetching from the table. To avoid hitting the table too often, spaces, tabs, and carriage returns, and line feeds are currently handled separately.

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
It's probably faster to actually check the common case of spaces, tabs, and newlines, but the current draft always does the lookup.

@typescript-bot perf test this

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this

**Comment by DanielRosenwasser** [comment]
I'm actually not sure if that's much better or not, it could just be a better run.

@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
@typescript test it
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
@typescript-bot test top400
@typescript-bot perf test this

**Comment by DanielRosenwasser** [comment]
@dragomirtitian maybe this balances out the parse-time hit from #58928. 😄

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this

File: src/compiler/scanner.ts [hunk 1]
```diff
@@ -1882,69 +1996,115 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
...
+                        tokenFlags |= TokenFlags.OctalSpecifier;
+                        return token = checkBigIntSuffix();
                     }
+                }
+
+                return token = scanNumber();
+            }
+
+            if (!(tokenCategory & TokenCategory.RecognizedMisc)) Debug.fail(`Unhandled token category ${tokenCategory}`);
+            switch (ch) {
```
Comment (rbuckton) [inline src/compiler/scanner.ts:2107]: I wonder if the `switch` here is optimized in any way or if it is doing case-by-case comparisons against each character code. If the latter, maybe it would be worthwhile to analyze what that most common character codes in JS are and change the order of the cases so that the more frequent cases are earlier in the list. I expect the perf improvement here actually comes from handling digit and identifier characters earlier than the `switch`, since those are the most frequent characters in a given file. If that's the case, some of the complexity and memory overhead could be reduced by just optimizing for those two cases.

File: src/compiler/scanner.ts [hunk 2]
```diff
@@ -1882,69 +1996,115 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
...
+
+            if (tokenCategory & TokenCategory.LineBreak) {
+                tokenFlags |= TokenFlags.PrecedingLineBreak;
+                pos += charSize(ch);
+                continue;
+            }
+
+            if (tokenCategory & TokenCategory.SimpleToken) {
+                pos++;
+                return token = tokenCategory & TokenCategory.SimpleTokenMask;
```
Comment (rbuckton) [inline src/compiler/scanner.ts:2063]: A comment here would be good for future readers as it isn't obvious at first glance how a `TokenCategory` becomes a token.

File: src/compiler/scanner.ts (hunk 1 above)
Comment (DanielRosenwasser) [reply src/compiler/scanner.ts:2107]: I think it is typically case-by-case comparisons. You're absolutely right that doing the checks earlier on for common classes; however, I think that checking for all the identifier starts isn't practical (or requires the same sort of fast-path/slow-path separation where we check for ranges), so that's why the token categorization is helpful here. By dismissing all the other types of categories, we can immediately launch into the fast path.

File: src/compiler/scanner.ts [hunk 3]
```diff
@@ -301,6 +301,120 @@ const regExpFlagToFirstAvailableLanguageVersion = new Map<RegularExpressionFlags
...
+        [CharacterCodes.greaterThan, TokenCategory.RecognizedMisc],
+        [CharacterCodes.question, TokenCategory.RecognizedMisc],
+        [CharacterCodes.caret, TokenCategory.RecognizedMisc],
+        [CharacterCodes.bar, TokenCategory.RecognizedMisc],
+        [CharacterCodes.backslash, TokenCategory.RecognizedMisc],
+        [CharacterCodes.hash, TokenCategory.RecognizedMisc],
+        [CharacterCodes.replacementCharacter, TokenCategory.RecognizedMisc],
+    ]
+) {
+    if (key < tokenCategoryLookup.length) {
```
[the remark this answers is not in the document]
Comment (DanielRosenwasser) [reply src/compiler/scanner.ts:410]: Can you be more specific? In the first two commits, I split this out from a map to an array and map to avoid the map lookup.

[the remark this answers is not in the document]
Comment (rbuckton) [reply src/compiler/scanner.ts:410]: I mean that, since we know ahead of time which characters belong in the `Map`, we can just construct the map with those character codes rather than looping and comparing. It's not that important, and if it makes more sense to have them visually grouped together, that's perfectly fine.

File: src/compiler/scanner.ts [hunk 4]
```diff
@@ -1882,69 +1996,115 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
...
+                        if (!tokenValue) {
+                            error(Diagnostics.Octal_digit_expected);
+                            tokenValue = "0";
                         }
-                        return token = SyntaxKind.WhitespaceTrivia;
+                        tokenValue = "0o" + tokenValue;
+                        tokenFlags |= TokenFlags.OctalSpecifier;
+                        return token = checkBigIntSuffix();
                     }
+                }
```
Comment (dmichon-msft) [inline src/compiler/scanner.ts:2101]:
```suggestion
                if (ch === CharacterCodes._0 && pos + 2 < end) {
                    switch (charCodeUnchecked(pos + 1)) {
                    case CharacterCodes.X:
                    case CharacterCodes.x:
                        pos += 2;
                        tokenValue = scanMinimumNumberOfHexDigits(1, /*canHaveSeparators*/ true);
                        if (!tokenValue) {
                            error(Diagnostics.Hexadecimal_digit_expected);
                            tokenValue = "0";
                        }
                        tokenValue = "0x" + tokenValue;
                        tokenFlags |= TokenFlags.HexSpecifier;
                        return token = checkBigIntSuffix();
                    case CharacterCodes.B:
                    case CharacterCodes.b:
                        pos += 2;
                        tokenValue = scanBinaryOrOctalDigits(/* base */ 2);
                        if (!tokenValue) {
                            error(Diagnostics.Binary_digit_expected);
                            tokenValue = "0";
                        }
                        tokenValue = "0b" + tokenValue;
                        tokenFlags |= TokenFlags.BinarySpecifier;
                        return token = checkBigIntSuffix();
                    case CharacterCodes.O:
                    case CharacterCodes.o:
                        pos += 2;
                        tokenValue = scanBinaryOrOctalDigits(/* base */ 8);
                        if (!tokenValue) {
                            error(Diagnostics.Octal_digit_expected);
                            tokenValue = "0";
                        }
                        tokenValue = "0o" + tokenValue;
                        tokenFlags |= TokenFlags.OctalSpecifier;
                        return token = checkBigIntSuffix();
                }
```

File: src/compiler/scanner.ts (hunk 1 above)
Comment (dmichon-msft) [reply src/compiler/scanner.ts:2107]: Assuming `CharacterCodes` is a `const enum`, the VM should be able to optimize the switch case quite heavily as a branch table, but that's a question for V8.

Comment (rbuckton) [reply src/compiler/scanner.ts:2107]: If the majority of identifier/keyword characters are ASCII characters, then it's fairly trivial to check for those major cases without an array or `Map` using a simple comparison:

```ts
if (
    (ch | 32) >= CharacterCodes.a && (ch | 32) <= CharacterCodes.z ||
    ch === CharacterCodes._ ||
    ch === CharacterCodes.$
) { /* [A-Za-z_$] */ }
```

If the array/map lookups are faster than this for the identifier case, then by all means keep it as is. I only point this out as you explicitly avoid hitting the table for line terminators and spaces. If that was an important optimization, then it might make sense to do the same kind of optimization for the most frequent (and easy to optimize for) characters in a source file, e.g., ASCII letters.

Comment (rbuckton) [reply src/compiler/scanner.ts:2107]: > Assuming `CharacterCodes` is a `const enum`, the VM should be able to optimize the switch case quite heavily as a branch table, but that's a question for V8.

We've generally found that not to be the case, which is why we've had to implement our own jump tables for functions like `forEachChild`. I'm not sure what heuristics V8 uses for `switch` optimization, if any.

File: src/compiler/scanner.ts (hunk 4 above)
Comment (DanielRosenwasser) [reply src/compiler/scanner.ts:2101]: It's hard to parse out what the suggestion is - can you explain it?

File: src/compiler/scanner.ts (hunk 1 above)
Comment (DanielRosenwasser) [reply src/compiler/scanner.ts:2107]: > If the majority of identifier/keyword characters are ASCII characters, then it's fairly trivial to check for those major cases without an array or Map using a simple comparison:

Fair, and this is what @ahejlsberg also suggested. The same way whitespace is prioritized, ASCII range is also very likely and worth prioritizing as a quicker up-front case instead of hitting the lookup table. It would mean there's some code duplication but maybe it's fine.

File: src/compiler/scanner.ts (hunk 4 above)
Comment (dmichon-msft) [reply src/compiler/scanner.ts:2101]: Ah, that's fair, I don't know why GitHub insists on showing the old code in the diff rather than diffing against your PR version of the code.

Suggestion is to promote the `pos + 2 < end` check to the same `if` that checks for `CharacterCodes._0`, since all branches beneath it use that condition, then replace the nested if statements with a single `switch (charCodeUnchecked(pos + 1))` with grouped case statements for each of the radix markers.

File: src/compiler/scanner.ts (hunk 1 above)
Comment (DanielRosenwasser) [reply src/compiler/scanner.ts:2107]: >> Assuming `CharacterCodes` is a `const enum`, the VM should be able to optimize the switch case quite heavily as a branch table, but that's a question for V8.

> We've generally found that not to be the case, which is why we've had to implement our own jump tables for functions like `forEachChild`. I'm not sure what heuristics V8 uses for `switch` optimization, if any.

Especially in the case where these character codes are so sparse - I am very doubtful that V8 would do such an optimization, and at best I'd guess we'd end up with something like a binary search.

**Comment by DanielRosenwasser** [comment]
#59244 tries to just check for identifiers up-front, and moves checks for `space` earlier. It is faster, but not as fast as the lookup approach here.
